packages/swingset-liveslots/src/virtualObjectManager.js

```diff
@@ -15,137 +15,273 @@ const { details: X, quote: q } = assert;
+    syscall.vatstoreSet(`vom.${baseRef}`, rawState);
+  };
+  /** @type {(collectionID: string) => void} */
+  const deleteBacking = baseRef => syscall.vatstoreDelete(`vom.${baseRef}`);
+  return makeCache(readBacking, writeBacking, deleteBacking);
+}
+
+function makeContextCache(makeState, makeContext) {
+  // non-writeback cache for "context" objects, { state, self/facets }
+  const readBacking = baseRef => {
+    const state = makeState(baseRef);
+    const context = makeContext(baseRef, state);
+    return context;
+  };
+  const writeBacking = _baseRef => undefined; // never called
+  const deleteBacking = _baseRef => undefined; // never called
+  return makeCache(readBacking, writeBacking, deleteBacking);
+}
+
+/**
+ * @typedef {import('@agoric/store/src/patterns/exo-tools.js').ContextProvider } ContextProvider
+ */
+
+/**
+ *
+ * @param {*} contextCache
+ * @param {*} getSlotForVal
+ * @returns {ContextProvider}
+ */
+const makeContextProvider = (contextCache, getSlotForVal) => {
+  return harden(rep => contextCache.get(getSlotForVal(rep)));
+};
+
+const makeContextProviderKit = (contextCache, getSlotForVal, facetNames) => {
+  /** @type { Record<string, any> } */
+  const contextProviderKit = {};
+  for (const [index, name] of facetNames.entries()) {
+    contextProviderKit[name] = rep => {
+      const vref = getSlotForVal(rep);
+      const { baseRef, facet } = parseVatSlot(vref);
+
+      // Without this check, an attacker (with access to both cohort1.facetA
+      // and cohort2.facetB) could effectively forge access to cohort1.facetB
+      // and cohort2.facetA. They could not forge the identity of those two
+      // objects, but they could invoke all their equivalent methods, by using
+      // e.g. cohort1.facetA.foo.apply(cohort2.facetB, [...args])
+      Number(facet) === index || Fail`illegal cross-facet access`;
+
+      return harden(contextCache.get(baseRef));
+    };
+  }
+  return harden(contextProviderKit);
+};
+
+function checkAndUpdateFacetiousness(
+  tag,
+  desc,
+  facetNames,
+  saveDurableKindDescriptor,
+) {
+  // The first time a durable kind gets a definition, the saved descriptor
+  // will have neither ".unfaceted" nor ".facets", and we must record the
+  // initial details in the descriptor.
+
+  if (!desc.unfaceted && !desc.facets) {
+    if (facetNames) {
+      desc.facets = facetNames;
+    } else {
+      desc.unfaceted = true;
+    }
+    saveDurableKindDescriptor(desc);
+    return;
+  }
+
+  // When a later incarnation redefines the behavior, it must match.
+
+  if (facetNames && desc.unfaceted) {
+    Fail`defineDurableKindMulti called for unfaceted KindHandle ${tag}`;
+  }
+  if (!facetNames && desc.facets) {
+    Fail`defineDurableKind called for faceted KindHandle ${tag}`;
+  }
+  const m = `durable kind "${tag}" facets don't match original definition`;
+  if (facetNames) {
+    for (const [idx, facet] of facetNames.entries()) {
+      assert(facet === desc.facets[idx], m);
+    }
+  }
```

Comment:
Suggestion for use of `Fail` and a more informative message:
```suggestion
  if (facetNames) {
    for (const [idx, facet] of facetNames.entries()) {
      facet === desc.facets[idx] ||
        Fail`durable kind ${q(tag)} facets don't match original definition at index ${q(idx)}`;
    }
  }
```

Reply:
If we aren't telling them both the original and the new list, then I don't think telling them the index is going to help much. And that `Fail` gets split by prettier into three ugly lines (it wants to avoid long lines, but the only legal place to split them is inside the quasiliteral holes, and it chooses the spot between `${(` and `tag`, it's gross).

I'll rewrite this to complain at the end of the comparison, and announce both lists.